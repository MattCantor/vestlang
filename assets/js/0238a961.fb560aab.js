"use strict";(self.webpackChunk_vestlang_docs=self.webpackChunk_vestlang_docs||[]).push([[133],{6336:(t,e,n)=>{n.d(e,{R:()=>r,x:()=>a});var s=n(9430);const i={},o=s.createContext(i);function r(t){const e=s.useContext(o);return s.useMemo(function(){return"function"==typeof t?t(e):{...e,...t}},[e,t])}function a(t){let e;return e=t.disableParentContext?"function"==typeof t.components?t.components(i):t.components||i:r(t.components),s.createElement(o.Provider,{value:e},t.children)}},6649:(t,e,n)=>{n.r(e),n.d(e,{assets:()=>c,contentTitle:()=>a,default:()=>u,frontMatter:()=>r,metadata:()=>s,toc:()=>d});const s=JSON.parse('{"id":"design/decisions","title":"decisions","description":"","source":"@site/docs/design/decisions.mdx","sourceDirName":"design","slug":"/design/decisions","permalink":"/vestlang/design/decisions","draft":false,"unlisted":false,"editUrl":"https://github.com/MattCantor/vestlang/docs/design/decisions.mdx","tags":[],"version":"current","frontMatter":{},"sidebar":"tutorialSidebar","next":{"title":"installation","permalink":"/vestlang/getting-started/installation"}}');var i=n(5270),o=n(6336);const r={},a=void 0,c={},d=[];function l(t){return(0,i.jsx)(i.Fragment,{})}function u(t={}){const{wrapper:e}={...(0,o.R)(),...t.components};return e?(0,i.jsx)(e,{...t,children:(0,i.jsx)(l,{...t})}):l()}}}]);