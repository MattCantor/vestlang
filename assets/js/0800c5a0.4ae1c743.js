"use strict";(self.webpackChunk_vestlang_docs=self.webpackChunk_vestlang_docs||[]).push([[412],{5368:(e,t,n)=>{n.r(t),n.d(t,{assets:()=>i,contentTitle:()=>a,default:()=>p,frontMatter:()=>c,metadata:()=>s,toc:()=>u});const s=JSON.parse('{"id":"Source/types","title":"types","description":"","source":"@site/docs/03-Source/types.mdx","sourceDirName":"03-Source","slug":"/Source/types","permalink":"/vestlang/Source/types","draft":false,"unlisted":false,"editUrl":"https://github.com/MattCantor/vestlang/docs/03-Source/types.mdx","tags":[],"version":"current","frontMatter":{},"sidebar":"tutorialSidebar","previous":{"title":"normalizer","permalink":"/vestlang/Source/normalizer"},"next":{"title":"decisions","permalink":"/vestlang/Design/decisions"}}');var o=n(5270),r=n(6336);const c={},a=void 0,i={},u=[];function l(e){return(0,o.jsx)(o.Fragment,{})}function p(e={}){const{wrapper:t}={...(0,r.R)(),...e.components};return t?(0,o.jsx)(t,{...e,children:(0,o.jsx)(l,{...e})}):l()}},6336:(e,t,n)=>{n.d(t,{R:()=>c,x:()=>a});var s=n(9430);const o={},r=s.createContext(o);function c(e){const t=s.useContext(r);return s.useMemo(function(){return"function"==typeof e?e(t):{...t,...e}},[t,e])}function a(e){let t;return t=e.disableParentContext?"function"==typeof e.components?e.components(o):e.components||o:c(e.components),s.createElement(r.Provider,{value:t},e.children)}}}]);