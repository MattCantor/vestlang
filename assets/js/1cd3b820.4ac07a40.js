"use strict";(self.webpackChunk_vestlang_docs=self.webpackChunk_vestlang_docs||[]).push([[344],{733:(e,t,r)=>{r.r(t),r.d(t,{assets:()=>i,contentTitle:()=>c,default:()=>m,frontMatter:()=>s,metadata:()=>n,toc:()=>u});const n=JSON.parse('{"id":"Source/grammar","title":"grammar","description":"","source":"@site/docs/03-Source/grammar.mdx","sourceDirName":"03-Source","slug":"/Source/grammar","permalink":"/vestlang/Source/grammar","draft":false,"unlisted":false,"editUrl":"https://github.com/MattCantor/vestlang/docs/03-Source/grammar.mdx","tags":[],"version":"current","frontMatter":{},"sidebar":"tutorialSidebar","previous":{"title":"Two-Tier Vesting","permalink":"/vestlang/References/Advanced/two-tier-vesting"},"next":{"title":"normalizer","permalink":"/vestlang/Source/normalizer"}}');var o=r(5270),a=r(6336);const s={},c=void 0,i={},u=[];function l(e){return(0,o.jsx)(o.Fragment,{})}function m(e={}){const{wrapper:t}={...(0,a.R)(),...e.components};return t?(0,o.jsx)(t,{...e,children:(0,o.jsx)(l,{...e})}):l()}},6336:(e,t,r)=>{r.d(t,{R:()=>s,x:()=>c});var n=r(9430);const o={},a=n.createContext(o);function s(e){const t=n.useContext(a);return n.useMemo(function(){return"function"==typeof e?e(t):{...t,...e}},[t,e])}function c(e){let t;return t=e.disableParentContext?"function"==typeof e.components?e.components(o):e.components||o:s(e.components),n.createElement(a.Provider,{value:t},e.children)}}}]);