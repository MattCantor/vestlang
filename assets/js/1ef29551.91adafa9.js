"use strict";(self.webpackChunk_vestlang_docs=self.webpackChunk_vestlang_docs||[]).push([[961],{1374:(e,t,n)=>{n.r(t),n.d(t,{assets:()=>u,contentTitle:()=>a,default:()=>p,frontMatter:()=>c,metadata:()=>s,toc:()=>i});const s=JSON.parse('{"id":"source/types","title":"types","description":"","source":"@site/docs/source/types.mdx","sourceDirName":"source","slug":"/source/types","permalink":"/vestlang/source/types","draft":false,"unlisted":false,"editUrl":"https://github.com/MattCantor/vestlang/docs/source/types.mdx","tags":[],"version":"current","frontMatter":{},"sidebar":"tutorialSidebar","previous":{"title":"normalizer","permalink":"/vestlang/source/normalizer"}}');var r=n(5270),o=n(6336);const c={},a=void 0,u={},i=[];function l(e){return(0,r.jsx)(r.Fragment,{})}function p(e={}){const{wrapper:t}={...(0,o.R)(),...e.components};return t?(0,r.jsx)(t,{...e,children:(0,r.jsx)(l,{...e})}):l()}},6336:(e,t,n)=>{n.d(t,{R:()=>c,x:()=>a});var s=n(9430);const r={},o=s.createContext(r);function c(e){const t=s.useContext(o);return s.useMemo(function(){return"function"==typeof e?e(t):{...t,...e}},[t,e])}function a(e){let t;return t=e.disableParentContext?"function"==typeof e.components?e.components(r):e.components||r:c(e.components),s.createElement(o.Provider,{value:t},e.children)}}}]);