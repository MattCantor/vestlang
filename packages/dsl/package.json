{
  "name": "@vestlang/dsl",
  "version": "0.1.0",
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    }
  },
  "types": "./dist/index.d.ts",
  "scripts": {
    "build": "rimraf dist src/generated && peggy --format es --dts -o src/generated/grammar.js src/grammar/*.peggy && tsc -p tsconfig.build.json",
    "dev": "peggy --format es --dts -w -o src/generated/grammar.js src/grammar/*.peggy",
    "test": "vitest"
  },
  "devDependencies": {
    "peggy": "^5.0.6",
    "typescript": "^5.9.2",
    "rimraf": "^6.0.1"
  }
}
