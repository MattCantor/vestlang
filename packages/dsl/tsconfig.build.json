{
  "compilerOptions": {
    "target": "ES2022",
    "module": "ES2022",
    "moduleResolution": "Bundler",
    "rootDir": "src",
    "outDir": "dist",
    "declaration": true,
    "allowJs": true,
    "checkJs": false,
    "strict": true,
    "esModuleInterop": true,
    "skipLibCheck": true,
    "resolveJsonModule": true
  },
  "include": [
    "src/**/*.ts",
    "src/**/*.d.ts",
    "src/generated/**/*.js",
    "src/generated/**/*.d.ts"
  ]
}
