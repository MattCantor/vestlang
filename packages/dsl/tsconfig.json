{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "target": "ES2020",
    "module": "ESNext",
    "moduleResolution": "Bundler",
    "allowJs": true,
    "types": [
      "vitest/globals"
    ]
  },
  "include": [
    "src",
    "tests"
  ]
}
