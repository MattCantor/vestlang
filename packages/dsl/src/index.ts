import * as parser from "./generated/grammar.js";
export type * from "./generated/grammar.d.js";

import type { Program } from "@vestlang/types";

export function parse(input: string): Program {
  const stmt = parser.parse(input) as Program;
  return stmt;
}
