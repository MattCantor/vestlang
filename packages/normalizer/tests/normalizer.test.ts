import { describe, it, expect } from "vitest";
import { normalizeProgram } from "../src/normalizer/index.js";
import type {
  AndCondition,
  AtomCondition,
  ConstrainedVestingNode,
  LaterOfVestingNode,
  Program,
} from "@vestlang/types";
import { parse } from "@vestlang/dsl";

/* ------------------------
 * Helpers
 * ------------------------ */

function norm(src: string): Program {
  const ast = parse(src);
  return normalizeProgram(ast);
}

/** Throw error if the first statement is not a singleton */
function getSingleton(program: Program) {
  const stmt = program[0];
  if (!stmt || stmt.expr.type !== "SINGLETON") {
    throw new Error("Expected first statement to be a SINGLETON");
  }
  return stmt.expr;
}

/** Getter for vesting_start expression (bare/constrained or selector) */
function getVestingStartItems(p: Program) {
  const vesting_start = getSingleton(p).vesting_start;
  return vesting_start &&
    (vesting_start.type === "EARLIER_OF" || vesting_start.type === "LATER_OF")
    ? vesting_start.items
    : [];
}

/* ------------------------
 * Selectors
 * ------------------------ */

describe("selectors: flatten, dedupe, collapse", () => {
  it("flattens nested LATER OF and dedupes identical items", () => {
    const out = norm(`
    VEST FROM
      LATER OF(
        EVENT a + 1 day,
        LATER OF(
          EVENT a + 1 day,
          DATE 2025-01-01
        )
      )
    OVER 12 months
    EVERY 1 month
`);
    const items = getVestingStartItems(out);

    // Should flatten to two unique items after dedupe
    expect(items.length).toBe(2);

    // Ensure there is no nested selector remaining at top-level items
    expect(
      items.every((x: any) => x.type === "BARE" || x.type === "CONSTRAINED"),
    ).toBe(true);
  });

  it("collapses to singleton after dedupe when both items are identical", () => {
    const out = norm(`
      VEST FROM
        LATER OF(
          EVENT a,
          EVENT a
        )
    `);

    // After normalization, vesting_start should not be a selector anymore
    const vesting_start = getSingleton(out).vesting_start;
    expect(
      vesting_start.type === "LATER_OF" || vesting_start.type === "EARLIER_OF",
    ).toBe(false);
  });
});

/* ------------------------
 * Constraints
 * ------------------------ */

// Duplicate AND constraints collapse into a single ATOM
describe("constraints: dedupe + singleton collapse", () => {
  it("AND(BEFORE b, BEFORE b) -> ATOM(BEFORE B)", () => {
    const out = norm(`
      VEST FROM EVENT start
        BEFORE EVENT b + 2 days
        AND BEFORE EVENT b + 2 days
    `);

    const vesting_start = getSingleton(out).vesting_start;
    if (vesting_start.type !== "CONSTRAINED") {
      throw new Error("Expected vesting start to be constrainted");
    }
    const c = (vesting_start as ConstrainedVestingNode).constraints;

    // After dedupe, AND has 1 item -> collapses to that item (ATOM)
    expect(c.type).toBe("ATOM");
    expect((c as AtomCondition).constraint.type).toBe("BEFORE");
    expect((c as AtomCondition).constraint.base.type).toBe("BARE");
  });
});

/** nested AND(...) flattens to a single AND with 3 items (no dedupe). */
describe("constraints: flatten AND", () => {
  it("AND(BEFORE a, AND(BEFORE b, BEFORE c)) → AND(BEFORE a, BEFORE b, BEFORE c)", () => {
    const out = norm(`
      VEST FROM EVENT start
          AND(
            BEFORE EVENT a,
            AND(
              BEFORE EVENT b,
              BEFORE EVENT c
            )
          )
    `);

    const vs = getSingleton(out).vesting_start;
    expect(vs.type).toBe("CONSTRAINED");
    const c = (vs as ConstrainedVestingNode).constraints;
    expect(c.type).toBe("AND");
    expect((c as AndCondition).items.length).toBe(3);
    // Sanity: each item is an ATOM constraint
    expect((c as AndCondition).items.every((x) => x.type === "ATOM")).toBe(
      true,
    );
  });
});

/** LATER OF with nested LATER OF flattens and dedupes to 2 unique items. */
describe("selectors (vesting_start): flatten + dedupe", () => {
  it("LATER OF(a, LATER OF(a, date)) → LATER_OF(a, date)", () => {
    const out = norm(`
      VEST
        FROM LATER OF(
          EVENT a + 1 day,
          LATER OF(
            EVENT a + 1 day,
            DATE 2025-01-01
          )
        )
    `);

    const vs = getSingleton(out).vesting_start;
    expect(vs.type).toBe("LATER_OF");
    expect((vs as LaterOfVestingNode).items.length).toBe(2);
    // Top-level items are vesting nodes (selector was flattened)
    expect(
      (vs as LaterOfVestingNode).items.every(
        (x: any) => x.type === "BARE" || x.type === "CONSTRAINED",
      ),
    ).toBe(true);
  });
});

/** LATER OF(dup, dup) collapses to bare node. */
describe("selectors (vesting_start): singleton collapse", () => {
  it("LATER OF(m1, m1) → m1", () => {
    const out = norm(`
      VEST FROM LATER OF( EVENT m1, EVENT m1 )
    `);

    const vs = getSingleton(out).vesting_start;
    expect(vs.type === "LATER_OF" || vs.type === "EARLIER_OF").toBe(false);
    expect(vs.type === "BARE" || vs.type === "CONSTRAINED").toBe(true);
  });
});

/** Focus: EARLIER OF(date, same date) collapses to single node in cliff. */
describe("cliff: selector dedupe + collapse", () => {
  it("EARLIER OF(d, d) → d inside CLIFF", () => {
    const out = norm(`
      VEST FROM EVENT start
        OVER 12 months EVERY 1 month
        CLIFF EARLIER OF(
          DATE 2026-01-01,
          DATE 2026-01-01
        )
    `);

    const cliff = getSingleton(out).periodicity.cliff;
    expect(cliff).toBeTruthy();
    // After dedupe, selector collapses to node
    expect(cliff!.type === "EARLIER_OF" || cliff!.type === "LATER_OF").toBe(
      false,
    );
    expect(cliff!.type === "BARE" || cliff!.type === "CONSTRAINED").toBe(true);
  });
});

/* ------------------------
 * Idempotence
 * ------------------------ */

describe("normalize is idempotent", () => {
  it("normalize(normalize(parse(src))) === normalize(parse(src))", () => {
    const src = `
      VEST FROM
        LATER OF(
          EVENT x + 1 month,
          LATER OF( DATE 2025-01-01, DATE 2025-01-01 )
        )
      OVER 12 months
      EVERY 1 month
    `;

    const once = norm(src);
    const twice = normalizeProgram(once as Program);

    // Property check: this is the real idempotence assertion.
    expect(twice).toEqual(once);

    // Optional: snapshot for human-friendly regression diff.
    expect(once).toMatchSnapshot();
  });
});
