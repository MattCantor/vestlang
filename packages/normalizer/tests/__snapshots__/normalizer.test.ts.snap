// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`normalize is idempotent > normalize(normalize(parse(src))) === normalize(parse(src)) 1`] = `
[
  {
    "amount": {
      "denominator": 1,
      "numerator": 1,
      "type": "PORTION",
    },
    "expr": {
      "periodicity": {
        "length": 1,
        "occurrences": 12,
        "type": "MONTHS",
      },
      "type": "SINGLETON",
      "vesting_start": {
        "items": [
          {
            "base": {
              "type": "DATE",
              "value": "2025-01-01",
            },
            "offsets": [],
            "type": "BARE",
          },
          {
            "base": {
              "type": "EVENT",
              "value": "x",
            },
            "offsets": [
              {
                "sign": "PLUS",
                "type": "DURATION",
                "unit": "MONTHS",
                "value": 1,
              },
            ],
            "type": "BARE",
          },
        ],
        "type": "LATER_OF",
      },
    },
  },
]
`;
